### PR TITLE
Create commands.md

### LINKED ISSUE #57: Add docs for ##[] commands from scripts
e.g. `##[error]Error message`

### PR DESCRIPTION
**Pull request by damccorm**
✨ [Rendered](https://github.com/actions/toolkit/blob/commands/docs/commands.md) ✨ 

Added docs on :: commands. Fixes #57

### CODE REVIEW

File: docs/commands.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# :: Commands
+
+The [core toolkit package](https://github.com/actions/toolkit/tree/master/packages/core) offers a number of convenience functions for
+setting results, logging, registering secrets and exporting variables across actions. Sometimes, however, its useful to be able to do
+these things in a script or other tool.
+
+To allow this, we permit a special `::` syntax which, if logged to stdout, will allow the runner to interpret your commands.
```
Comment (JasonEtco) [inline docs/commands.md]: Tiny suggestion, making `stdout` look a little more technical:

```suggestion
To allow this, we permit a special `::` syntax which, if logged to `stdout`, will allow the runner to interpret your commands.
```

File: docs/commands.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+The following commands are all supported:
+
+### Set an environment variable
+
+To set an environment variable, use `::set-env`:
+
+`echo ::set-env name=FOO::BAR`
```
Comment (JasonEtco) [inline docs/commands.md]: For easier scanning and nicer formatting, what do you think about making these block-level code blocks?

`echo ::set-env name=FOO::BAR`

vs:

```
echo ::set-env name=FOO::BAR
```

File: docs/commands.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+Running `$PATH` in a future step will now return `BAR:{Previous Path}`;
+
+### Set outputs
+
+To set an output for the step, use `set-output`:
+
+`echo ::set-output name=FOO::BAR`
+
+Running `steps.[step-id].outputs.FOO` in your Yaml will now give you `BAR`
```
Comment (JasonEtco) [inline docs/commands.md]: I think this is a little unclear - maybe we can add some example usage?

```yaml
steps:
  - name: Set the value
    id: step_one
    run: echo ::set-output name=FOO::BAR
  - name: Use it
    run: echo ${{ steps.step_one.outputs.FOO }}
```

And/or link out to `core.setOutput`, which could also show a similar example.

**Comment by 1138-4EB** [comment]
Maybe `##[debug]` in https://github.com/actions/toolkit/blob/master/docs/action-debugging.md#step-debug-logs needs to be updated to `::[debug]`?

BTW, is this related to the `::[group]`/`::[endgroup]` feature? I'm currently using it to provide a 'clean' first view. See https://github.com/1138-4EB/ghdl/runs/210613754#step:3:46. I think it'd be useful if this was documented too. Moreover, it'd be really useful to:

- Allow nesting groups.
- Make timing a group a built-in feature. Ideally, show the elapsed time in the same row as the arrows (first/title line of the group), not in the end.
- Print the arrows (expand/collapse) in the margin, not in the body. And add vertical indentation lines.

I don't know if any of these features is available already through some other keyword.

File: docs/commands.md (hunk 1 above)
Comment (stephenmichaelf) [inline docs/commands.md]: Nit: permit -> provide?

Comment (stephenmichaelf) [inline docs/commands.md]: Interpret -> perform special behavior based on

File: docs/commands.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+To prepend a string to PATH, use `::addPath`:
+
+`echo ::add-path::BAR`
+
+Running `$PATH` in a future step will now return `BAR:{Previous Path}`;
+
+### Set outputs
+
+To set an output for the step, use `set-output`:
```
Comment (stephenmichaelf) [inline docs/commands.md]: Missing `::` at start of `set-output`

File: docs/commands.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+`echo ::set-output name=FOO::BAR`
+
+Running `steps.[step-id].outputs.FOO` in your Yaml will now give you `BAR`
+
+### Masking Values from Logs
```
Comment (stephenmichaelf) [inline docs/commands.md]: Nit: Masking values from logs -> Masking values in logs?

File: docs/commands.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+To mask a value from the logs, use `add-mask`:
```
Comment (stephenmichaelf) [inline docs/commands.md]: Same here, missing `::`. If we want this line to not have it, that works too. But we want to be consistent :) Right now it's mixed with some having it and some not.

File: docs/commands.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+`echo ::add-mask::BAR`
+
+Now, future logs containing BAR will be masked. E.g. running `echo "Hello FOO BAR World"` will now print `Hello FOO **** World`.
```
Comment (stephenmichaelf) [inline docs/commands.md]: Nit: `****` -> `***`?

Comment (stephenmichaelf) [reply docs/commands.md]: Or is it always a fixed amount to keep it more secret?

File: docs/commands.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+CAUTION: Do **not** mask short values if you can avoid it, it could render your output unreadable (and future steps' output as well).
+For example, if you mask the letter `l`, running `echo "Hello FOO BAR World"` will now print `He*********o FOO BAR Wor****d`
+
+### Log Level
+
+Finally, there are several commands to emit different levels of log output.
+To emit [debug](https://github.com/actions/toolkit/blob/master/docs/action-debugging.md) output, use `::debug`:
```
Comment (stephenmichaelf) [inline docs/commands.md]: This may be easier to understand as a table of log level - echo command.

File: docs/commands.md (hunk 7 above)
Comment (damccorm) [reply docs/commands.md]: I think its always fixed

File: docs/commands.md (hunk 8 above)
Comment (damccorm) [reply docs/commands.md]: I like it - lmk if my changes are what you were thinking

**Comment by damccorm** [comment]
Thanks for the feedback, incorporated it all in!

**Comment by JasonEtco** [comment]
@damccorm super excited for this. I have a potentially dumb question: is there a reason this is being added to the `actions/toolkit` docs rather than [the GitHub Actions developer docs](https://help.github.com/en/categories/automating-your-workflow-with-github-actions)? I ask because folks that don't want to write their actions in TS/JS may not come across it if they don't need the toolkit, and those that do would just use the `core` methods.

**Comment by damccorm** [comment]
@JasonEtco IMO it should probably end up in both places. In general, I like starting here since this is public and I can get more community feedback 😄

**Pull request merged**